### DISCUSSION TITLE
Setting up a reversing camera

### DISCUSSION THREAD

**Discussion by Geekyadz** [opencardev/crankshaft#659] [General]
I would like to setup a reversing camera using Crankhaft.

From looking at the Wiki, it seems to want to use a Raspberry Pi camra module (or one that uses the DSI connection)

I have also seen this issue: https://github.com/opencardev/crankshaft/issues/407

So here's my problem at the moment:

Just as a "proof of concept" I am trying to use the "use phone as a web camera" option that is on my Pixel 7a - this works with Discord and OBS on  my PC. I know, it's a different kettle of fish, but I would have thought that Crankshaft would also be able to detect it.

Admittidly I have to change the mode that the USB is in to change it, but I would have thought that once I have chnaged the momde it would be detected by Crankshaft as I can see that the phone disconnects then reconnects once in webcamera mode.

So far my testing today I have been using SSH into Crankshaft and it hasn't gave me much sucess; I enabled the raspberry pi camera in the raspi-config as instructed to in the Wiki but that hasn't helped; all I'm getting is a saying that the module cannot be detected (most likely cos I don't have one) and when I tried the crankshaft rearcam show command it didn't show me anything.

Obvioulsy I'm a little lost on this, and before I rush out buying multiple parts and finding things that don't work, I would rather have a look to see if I can get it woulding using the USB camera feature on my phone.

**Comment by Geekyadz** [comment]
Update on this: so after some digging and changing of plans it appears that there's something about the raspberry pi 4 that's different.

When commenting out some lines for a Pi 3 model B cam overlay detects what I'm now using and had an error but commenting out the same the same lines for the Pi 4 it gives me a different error which I also initially had on the Pi 3

**Comment by Geekyadz** [comment]
Update:

After posting on [Cam_Overlay](https://github.com/meekys/cam_overlay/issues/31) about getting MPEG or YUYV mode working for my capture card (gone with YUYV) and obtaining a Raspberry Pi Model B (though it does look like a Raspberry Pi 3 Model B+ - the box makes no mention of whether it is a B+ or not) I have now got it working, and now I just ned a couple of things done for my setup to finish it off.

Video of concept: https://youtu.be/QF2X8LUEzLY

I used a push button cos at this moment of time that's all I have.

To get it working, you will need the following hardware (sorry, these are United Kingdom links):

Pi 3 I brought: https://amzn.to/4cd0VpF

Capture card: https://amzn.to/3XzvEsw

AV to HDMI converter: https://amzn.to/4cd4tIq

HDMI to HDMI lead also required, but you can pick these up from virtually anywhere

Reversing camera - again, Halfords, Amazon, or even use Google to find a shop that's great for you.

Something with a AV output - in my case I am using a PlayStation 2.

Note: [buttons](https://github.com/opencardev/crankshaft/wiki/Custom-Buttons) must not be used - otherwise this will not work! I found this out the hard way.

The process:

Make sure Crankshaft has ran at least once, and enable debug mode for testing and connecting by SSH. Whilst you are at it, comment (ie put # in front of dtoverlay=vc4-fcms-v3d so it reads:

`#dtoverlay=vc4-fcms-v3d` in config.txt on the root of the SD card.

Edit `/boot/crankshaft/crankshaft_env.sh` and enable [debug mode](https://github.com/opencardev/crankshaft/wiki/Dev-Mode-and-Debug-Mode) and also set Crankshaft to boot using X11.

Once Crankshaft has booted, in Settings > CS HW > put USB-Cam to Enabled, and reboot.

Make sure everything is turned off, and connect the USB capture card (you might want an USB extension lead) to the Pi 3, and connect the HDMI cable to both the capture card and the AV2HDMI converter - power this up separately if you are short on USB ports from the Pi 3

Connect wire to [GPIO pin 26](https://pinout.xyz/pinout/pin37_gpio26/) and a ground wire to either a rocker switch or a push button if you want to test this out first - this is for triggering the rearcam mode.

Boot up the Pi 3.

SSH into the Pi and log in. The code is what to use into SSH

First, lets see if it's all being detected: `v4l2-ctl --list-devices`

![image](https://github.com/opencardev/crankshaft/assets/30429165/a030ffac-9522-4507-a731-b487299ebe0d)

Now we check for what formats the capture card are supported for putting it onto the Pi: `v4l2-ctl --list-formats`

![image](https://github.com/opencardev/crankshaft/assets/30429165/2e257941-7530-43e3-ba23-c076251d4755)

MPEG will cause a black screen unless you have the CPP branch of cam_overlay installed. YUYV is supported in the cam_overlay that is installed with Crankshaft.

So, to get it to switch: `v4l2-ctl --device /dev/video0 --set-fmt-video=pixelformat=YUYV`

![image](https://github.com/opencardev/crankshaft/assets/30429165/16748c64-a8b8-4250-b993-27909d8b7a76)

Ok, doesn't look like that done anything. Let's confirm with SSH first: `crankshaft rearcam show`

![image](https://github.com/opencardev/crankshaft/assets/30429165/e1d34c22-7547-4e31-9625-d867d6d2b191)

(note: on my Pi now that I have programmed up the rearcam button, it shows a blank screen then hides again - however pressing my button makes it work)

If you see something, then it sounds like you are good to go. So `crankshaft rearcam hide` to exit that on your Pi.

Setting up the Pin for rearcam triggering:

Remember setting up the wires from earlier? Good. Well, into settings on Crankshaft, select CS Env and for Rearcam Pin set it to 26.

Reboot Pi (disconnect SSH first)

Once rebooted, SSH and `v4l2-ctl --device /dev/video0 --set-fmt-video=pixelformat=YUYV` again.

Test the button: you should find that you need to hold it to see the picture.

When you ready, get a camera and test it.

To do: find a way of getting `v4l2-ctl --device /dev/video0 --set-fmt-video=pixelformat=YUYV` to run at startup

**Comment by HUG1N** [comment]
Thank you very much for this detailed instruction @Geekyadz, it worked very well for me.

I did encounter some problems though, similar to [this](https://github.com/meekys/cam_overlay/issues/3#issue-372281546) issue, and commenting out the `dtoverlay=vc4-fkms-v3d` command lead to an unsuccessful boot of the Raspberry Pi.
Instead of dealing with the putative OpenGL issue, updating crankshaft lead to success with the rearview camera in the way you described in your tutorial.
To be able to update, the system has to be remounted as writable by using `mount -o remount, rw /`  (maybe not the best/correct way to do it, but it worked for me).

I put the `v4l2-ctl --device /dev/video0 --set-fmt-video=pixelformat=YUYV` command in the **/crankshaft/startup.sh** file and it works like a charme for both the **USB webcam** (Vitade 62H) and the **HDMI capture card** (Newhope;  ASIN: ‎B09NQM16VY) I used.

**Reply by Geekyadz** [reply]
> Thank you very much for this detailed instruction @Geekyadz, it worked very well for me.
> 
> I did encounter some problems though, similar to [this](https://github.com/meekys/cam_overlay/issues/3#issue-372281546) issue, and commenting out the `dtoverlay=vc4-fkms-v3d` command lead to an unsuccessful boot of the Raspberry Pi.
> Instead of dealing with the putative OpenGL issue, updating crankshaft lead to success with the rearview camera in the way you described in your tutorial.
> To be able to update, the system has to be remounted as writable by using `mount -o remount, rw /`  (maybe not the best/correct way to do it, but it worked for me).
> 
> I put the `v4l2-ctl --device /dev/video0 --set-fmt-video=pixelformat=YUYV` command in the **/crankshaft/startup.sh** file and it works like a charme for both the **USB webcam** (Vitade 62H) and the **HDMI capture card** (Newhope;  ASIN: ‎B09NQM16VY) I used.
> 

Thankyou for responding and letting me know that what I had mentioned kinda made sense for you to get it working for you.

Can you please confirm what model of pi you're using? 

I should have put this as an update, but yes after a bit of teasing for me last week putting that like into startup.sh has selected the correct mode. 

I've also decided to use a latching switch for toggling/triggering of the USB camera; I was thinking that it's probably possible to get it working with using a push button switch to toggle it on and off, but then it's also having the time and patience to test it.

**Reply by HUG1N** [reply]
Sure, I am using the Pi 3B (also not B+). 

I am also using a latching switch as I had some laying around and it works very well.

**Reply by Geekyadz** [reply]
Interesting, I only got your initial error with a Pi 4B when commenting out the dtoverlay=vc4-fkms-v3d lines for it. 

Also how did you run updates?

**Reply by HUG1N** [reply]
I also encountered the boot error on a Pi 4B but due to the lower heat development and power consumption I stick to the Pi3 without further testing on the Pi4.

I activated the devmode (which did not show any effect to me) and remounted the system as writable with `mount -o remount, rw /` and run a simple `sudo apt-get update` and `sudo apt-get upgrade` which in total led to around 10 minutes of updating time. Without writing permission I got a "(30: Read-only file system)" error.

**Reply by Geekyadz** [reply]
The Pi 4B needs 5 volts at 3 amps. I've been using it for the past year or so. I also use this heat sink and fans on it: https://thepihut.com/products/dual-fan-heatsink-case-for-raspberry-pi-4?variant=20679594180670

The Pi 3B needs 5 volts 2.4 amps and when I can remember I'll post a link to the case I'm currently using.

As for the updates: yeah maybe I should run them on mine, but I haven't needed to to get it working for me.

**Reply by HUG1N** [reply]
Nah, if it works, don't update it. I also would've prefer to keep it as "original" as possible but I was unable to make other solutions work.

I probably will get one of these "full body" cooling solutions but for now, the Pi3 stays cool enough and has enough power to run the whole project smoothly. I plan to enclose the Pi and the touch screen in a box on my car's dashboard including a small fan so I will find out soon, if heat will become a problem.

**Reply by Geekyadz** [reply]
Here's the case I use for the Pi 3B: https://amzn.to/4bUykEF

It's a little tall for my liking (but we are talking only millimetres)

**Comment by georgi-m-iliev** [comment]
I want to express my gratitude for this guide, which helped me to get my USB capture card to work! I would like to ask some questions just to clarify the compatibility of cam_overlay and Crankshaft. I am using a Pi Zero 2 W, that should have the same hardware as a Pi 3 B(+), as far as I'm concerned.  Do you run Crankshaft in EGL or X11 mode? After reading everything that I could find on the topic, I am coming to the conclusion that cam_overlay works only in X11 mode. When I try EGL mode, I get the following message from cam_overlay `* failed to add service - already in use?`. Are you aware of any way to run cam_overlay in EGL mode? I, personally, prefer it because the graphics seem of better quality and boot time is a tad bit shorter. 

As a side note, I wasn't able to boot Crankshaft without dtoverlay=vc4-fkms-v3d so it is out of question to fix cam_overlay with removing it.

**Reply by PointyFinger** [reply]
I am having the same issue on an RPI4B. It doesn't work in either EGl or X11 mode. I get the same error * failed to add service - already in use? regardless of what I do. Any suggestions?
Commenting out dtoverlay=vc4-fkms-v3d is not feasible, as Crankshaft will not boot without it.

**Reply by Geekyadz** [reply]
> I am having the same issue on an RPI4B. It doesn't work in either EGl or X11 mode. I get the same error * failed to add service - already in use? regardless of what I do. Any suggestions? Commenting out dtoverlay=vc4-fkms-v3d is not feasible, as Crankshaft will not boot without it.

Which is why I "gave up" on the Raspberry 4B, bit the bullet and went for a Raspberry 3B instead

**Reply by Geekyadz** [reply]
> I want to express my gratitude for this guide, which helped me to get my USB capture card to work! I would like to ask some questions just to clarify the compatibility of cam_overlay and Crankshaft. I am using a Pi Zero 2 W, that should have the same hardware as a Pi 3 B(+), as far as I'm concerned. Do you run Crankshaft in EGL or X11 mode? After reading everything that I could find on the topic, I am coming to the conclusion that cam_overlay works only in X11 mode. When I try EGL mode, I get the following message from cam_overlay `* failed to add service - already in use?`. Are you aware of any way to run cam_overlay in EGL mode? I, personally, prefer it because the graphics seem of better quality and boot time is a tad bit shorter.
> 
> As a side note, I wasn't able to boot Crankshaft without dtoverlay=vc4-fkms-v3d so it is out of question to fix cam_overlay with removing it.

Currently? I cannot remember. Everytime I read this and say "I must check this" when I use my car/Raspberry Pi 3B and the amount of times I forget is more times than I currently care to mention, so you have my apologies instead.

I haven't got a Pi Zero 2 W so I'm afraid I can't answer that, though Wikipedia did advice that they are similar hardware to the Raspberry Pi 3B (or at least I read something from Wikipedia that suggested that.

I must have found something between the threads that I have linked in here and the WiKi for Crankshaft

**Reply by Geekyadz** [reply]
> After reading everything that I could find on the topic, I am coming to the conclusion that cam_overlay works only in X11 mode. When I try EGL mode, I get the following message from cam_overlay `* failed to add service - already in use?`. Are you aware of any way to run cam_overlay in EGL mode? I, personally, prefer it because the graphics seem of better quality and boot time is a tad bit shorter.

So having changed SD card and having to rsetup this, I have found that only X11 works for me, and checking my other SD card that was setup, yes it was booting into X11. I have also updated my guide.

**Comment by Stimpson-J-Cat** [comment]
After doing all this, I can see my device and the formats but when I 'crankshaft rearcam show', I just get a prompt, unlike yours, where it tells you what it is opening and at what dimensions. I am using a 3B with the official RPi touch screen. I have USB cam enabled, rearcam pin set to 26 and have x11 mode enabled. 

Any ideas?  What have I missed?

**Reply by Geekyadz** [reply]
Well, what are you trying to use? 

What modes are supported by the camera? 

Can't help you if you don't tell me everything

**Reply by Stimpson-J-Cat** [reply]
YUYV is supported, but I did try both. When I tried the MPEG setting, it didn't even get as far and I never got the cursor back at the prompt.

**Reply by Geekyadz** [reply]
> YUYV is supported, but I did try both. When I tried the MPEG setting, it didn't even get as far and I never got the cursor back at the prompt.

So what happens when you use the pin, assuming you have wired a switch correctly? 
I recall on mine it seemed to stop allowing me to do the show rear cam thing on mine once I setup the switch but when I used the switch it's fine.

I still need to get a camera for mine.
